reconcilers/childset.go

```diff
@@ -0,0 +1,364 @@
+	// or nil if no children should exist. Each resource returned from this method must be claimed
+	// by the OurChild method with a stable, unique identifier returned. The identifier is used to
+	// correlate desired and actual child resources.
+	//
+	// To skip reconciliation of the child resources while still reflecting an existing child's
+	// status on the reconciled resource, return OnlyReconcileChildStatus as an error.
+	DesiredChildren func(ctx context.Context, resource Type) ([]ChildType, error)
+
+	// ReflectChildrenStatusOnParent updates the reconciled resource's status with values from the
+	// child reconciliations. Select types of errors are captured, including:
+	//   - apierrs.IsAlreadyExists
+	//
+	// Most errors are returned directly, skipping this method. The set of handled error types
+	// may grow, implementations should be defensive rather than assuming the error type.
+	ReflectChildrenStatusOnParent func(ctx context.Context, parent Type, result ChildSetResult[ChildType])
+
+	// HarmonizeImmutableFields allows fields that are immutable on the current
+	// object to be copied to the desired object in order to avoid creating
+	// updates which are guaranteed to fail.
+	//
+	// +optional
+	HarmonizeImmutableFields func(current, desired ChildType)
+
+	// MergeBeforeUpdate copies desired fields on to the current object before
+	// calling update. Typically fields to copy are the Spec, Labels and
+	// Annotations.
+	MergeBeforeUpdate func(current, desired ChildType)
+
+	// ListOptions allows custom options to be use when listing potential child resources. Each
+	// resource retrieved as part of the listing is confirmed via OurChild.
+	//
+	// Defaults to filtering by the reconciled resource's namespace:
+	//     []client.ListOption{
+	//         client.InNamespace(resource.GetNamespace()),
+	//     }
+	//
+	// +optional
+	ListOptions func(ctx context.Context, resource Type) []client.ListOption
+
+	// OurChild is used when there are multiple ChildReconciler for the same ChildType controlled
+	// by the same reconciled resource. The function return true for child resources managed by
+	// this ChildReconciler. Objects returned from the DesiredChildren function should match this
+	// function, otherwise they may be orphaned. If not specified, all children match. Matched
+	// child resources must also be uniquely identifiable with the IdentifyChild method.
+	//
+	// OurChild is required when a Finalizer is defined or SkipOwnerReference is true.
+	//
+	// +optional
+	OurChild func(resource Type, child ChildType) bool
+
+	// IdentifyChild returns a stable identifier for the child resource. The identifier is used to
+	// correlate desired child resources with actual child resources. The same value must be returned
+	// for an object both before and after it is created on the API server.
+	//
+	// Non-deterministic IDs will result in the rapid deletion and creation of child resources.
+	IdentifyChild func(child ChildType) string
```

Comment:
Would returning `error` be helpful here? In case users want to use that error to update parent status.

Reply:
This method is only called for resources that are either returned from DesiredChildren or actual resources that passed OurChild. It should always be able to produce an id. Returning empty string will result in the resource being deleted.

Covered by this test case https://github.com/vmware-labs/reconciler-runtime/pull/396/files#diff-90217b810ea3e7bc2b12f6613f8c7e88c01bbf44c159890c73426907750cd371R414